### PR TITLE
Uncomplicated Customr Role and Respawn Timer Information

### PR DESCRIPTION
**Pull request by warquys**
ref to this post https://discord.com/channels/1170301876990914631/1226555363998826547/1226555363998826547

### CODE REVIEW

**Comment by warquys** [comment]
Two choices are possible. Be activating the patch only when RespawnTimer is detected or making two different versions using preprocessor directives.

**Comment by FoxWorn3365** [comment]
We can use reflection, I've used it for the UncomplicatedCustomItems support

**Comment by warquys** [comment]
Detecting will avoid having to download the dll if the plugin is installed later for Owners. However this creates unrecognized symbols and can cause errors when attributes are inspected and RespawnTimer is not there. So yea this great if noboy do relexion on the class use for patch.

**Comment by FoxWorn3365** [comment]
Never had this problems, we just need to handle well the Assembly and the detection

**Comment by FoxWorn3365** [comment]
I'll review this as soon as I'll get home

File: UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj [hunk 1]
```diff
@@ -462,12 +462,19 @@
     <Reference Include="zxing, Version=0.16.9.0, Culture=neutral, PublicKeyToken=4e88037ac681fe60, processorArchitecture=MSIL">
       <SpecificVersion>False</SpecificVersion>
       <HintPath>$(EXILED_REFERENCES)\zxing.dll</HintPath>
+    </Reference>
+	<Reference Include="RespawnTimer, Version=0.0.0.0, Culture=neutral, Culture=neutral, processorArchitecture=MSIL">
```
Comment (FoxWorn3365) [inline UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj:466]: we can't do that as the plugin will require the `RespawnTimer.dll` in order to work

File: UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj [hunk 2]
```diff
@@ -36,17 +38,14 @@
...
     <ErrorReport>prompt</ErrorReport>
-    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
     <OutputPath>bin\x64\Release\</OutputPath>
     <DefineConstants>TRACE</DefineConstants>
     <Optimize>true</Optimize>
     <DebugType>pdbonly</DebugType>
     <PlatformTarget>x64</PlatformTarget>
-    <LangVersion>11</LangVersion>
```
Comment (FoxWorn3365) [inline UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj:49]: why?

File: UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Reflection.Emit;
+using Exiled.API.Features;
+using Exiled.API.Features.Roles;
+using HarmonyLib;
+using PlayerRoles;
+using Respawning;
+using RespawnTimer.API.Features;
```
Comment (FoxWorn3365) [inline UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs:10]: The whole file requires RespawnTimer as hard-dependency, we must use reflection

**Review by FoxWorn3365 — CHANGES_REQUESTED** [review]
We should probably use - or create - the APIs of RespawnTimer instead of using Harmony

File: UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj (hunk 1 above)
Comment (warquys) [reply UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj:466]: Yea this is why i put it in an other plugin.

File: UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj (hunk 2 above)
Comment (warquys) [reply UncomplicatedCustomRoles/UncomplicatedCustomRoles.csproj:49]: The Lange version get moved.
https://github.com/UncomplicatedCustomServer/UncomplicatedCustomRoles/pull/20/files#diff-2e672f6a45c0ce56d40bd884bf5263c3c93fe1648a4dae598026165f629f0cc4R13

File: UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs (hunk 3 above)
Comment (warquys) [reply UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs:10]: This is why I used an external plugin to avoid symbols not defined in the main plugin.

Comment (FoxWorn3365) [reply UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs:10]: reflection does not causes that, see https://github.com/UncomplicatedCustomServer/UncomplicatedCustomRoles/blob/main/UncomplicatedCustomRoles/Manager/SpawnManager.cs#L166

Comment (warquys) [reply UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs:10]: It is better to have a compilation error than a RunTime error. When we use reflection the compiler does not know if the method continues to exist

Comment (warquys) [reply UncomplicatedCustomRoles/Patch/RespawnTimerPatch.cs:10]: NB: i will change it with relfxion. I had a meeting today. And this afternoon I am doing practical work. I should also correct and put mark on some projects soon. Just to say "yea i ave some work... I back soon to do the edit".

**Comment by FoxWorn3365** [comment]
I saw that with reflection you get the Dictionary<string, Func<Player, string>> and save it in the plugin but because it's reflection you don't create a "pointer" but a copy of the Dictionary so editing it won't cause the edit on the reflected plugin (far I can remember).

Another thing is the param in the config for the compatibility: we can remove that because in order to get the role shown users have to put the {role} in the RespawnTimer config.

I've created a method to add a Func<Player, string> with a string in the TimerView file so you can use a method (more easy with reflection).

Oh and avoid putting functions inside the Plugin class, feel free to create a new file for thie features inside Manager/

**Comment by warquys** [comment]
No when you create a variable for a class in C#, this varaible is a pointer. When i get the value by reflexion i call the Get function of the property. It return the pointer of the class stored in the back field of the property.
I do not see the need to `AddReplacer` an function aliassing the `Add` of the dictionary.  ReplaceHelper is public...

**Review by FoxWorn3365 — COMMENTED** [review]
the code looks good, have you tested it?

**Comment by warquys** [comment]
Yes i tested it for each commit.

**Comment by FoxWorn3365** [comment]
this will be merged when the dev-main branch will be merged into main.
Next time plz do the pr on `dev-main`

**Review by FoxWorn3365 — APPROVED** [review]

**Pull request merged**